client/src/hooks/useMessenger.js

```diff
@@ -0,0 +1,281 @@
+    makeTempMessage(content);
+    try {
+      const response = await fetch(
+        `/api/convos/${currentConvo._id}/messages/`,
+        {
+          method: "post",
+          body: JSON.stringify({ content }),
+          headers: {
+            "Content-Type": "application/json",
+          },
+        }
+      );
+      if (response.ok) {
+        socket.emit("message", {
+          conversation: currentConvo,
+          author: auth.user,
+          content,
+        });
+      } else {
+        createSnack({
+          message: "Message unable to send",
+          severity: "error",
+        });
+        // Should add some message flag component or just remove the message from the messages array.
+      }
+    } catch (err) {
+      console.log(err);
+    }
+  };
+
+  // Constructs temp message to add to conversation prior to API call.
+  const makeTempMessage = (content, author = auth.user) => {
+    const message = {
+      dateCreated: Date.now(),
+      conversation: currentConvo,
+      author,
+      content,
+    };
+    messages
+      ? setMessages((currentMessages) => [...currentMessages, message])
+      : setMessages([message]);
+    updateConversation(message);
+  };
+
+  // socket listeners
+  useEffect(() => {
+    // Sorts convos based on online status
+    const sortConvos = (convos) => {
+      return convos.sort((c) => (c.isOnline ? -1 : 1));
+    };
+
+    // maps inital array of connected users
+    const setOnlineStatus = (users) => {
+      return allConvos.map((c) =>
+        users.some((user) => user._id === c.users[0]._id)
+          ? { ...c, isOnline: true }
+          : c
+      );
+    };
+
+    // maps single user
+    const mapUserStatus = (user, online) => {
+      return allConvos.map((c) =>
+        c.users[0]._id === user._id ? { ...c, isOnline: online } : c
+      );
+    };
+
+    // Handles new messages based on currentConvo
+    socket.on("newMessage", (message) => {
+      message.conversation._id === currentConvo._id
+        ? makeTempMessage(message.content, message.author)
+        : updateConversation(message, message.conversation);
+    });
+
+    socket.on("conversation", (conversation) => {
+      setAllConvos((currentConvos) => [...currentConvos, conversation]);
+      socket.emit("users");
+    });
+
+    socket.on("users", (payload) => {
+      let users = payload.map((item) => item.user);
+      setAllConvos(sortConvos(setOnlineStatus(users)));
+    });
+
+    socket.on("user connected", (user) => {
+      setAllConvos(sortConvos(mapUserStatus(user, true)));
+    });
+
+    socket.on("user disconnected", (user) => {
+      setAllConvos(sortConvos(mapUserStatus(user, false)));
+    });
+
+    return () => {
+      // Need to find some better way to do this
+      socket.off("newMessage");
+      socket.off("users");
+      socket.off("conversation");
+      socket.off("user connected");
+      socket.off("user disconnected");
+    };
+  }, [allConvos, currentConvo]);
```

Comment:
Ideally you wouldn't want this running each time the conversations update. Since the dependencies listed don't change what happens in the `useEffect`, it'd be better to remove them.

To suppress the linter error, you can place this on the line before the one where you list the dependencies:

```
// eslint-disable-line react-hooks/exhaustive-deps
```